tla-io/src/main/scala/at/forsyte/apalache/io/annotations/parser/CommentPreprocessor.scala

```diff
@@ -0,0 +1,258 @@
+package at.forsyte.apalache.io.annotations.parser
+
+/**
+ * <p>This class preprocesses comments as they are produced by SANY and does the following:</p>
+ *
+ * <ul>
+ * <li>Unconditionally remove the comment tokens: (*, *), and \*.</li>
+ * <li>Extract candidates for annotations.</li>
+ * <li>Exclude candidates for annotations in double comments, e.g., (* (* ... *) *).</li>
+ * <li>Preserve the text outside the annotations and double comments.
+ * </ul>
+ *
+ * <p>The main design principle of the preprocessor is that it should not fail on any input.
+ * The preprocessor should respect the structure of the comments, but it may return ill-formed
+ * annotations, if they are ill-formed in the original text. This preprocessor prunes the comment
+ * tokens, even in arguably valid cases (e.g., inside a string). Although it is usually not a good idea
+ * to write an ad-hoc lexer, it is not obvious to me, how to use a lexer generator here.
+ * However, if someone knows how to write this preprocessor with a lexer generator, it would be great!
```

Comment:
Did you look at some existing lexer generators for Scala? If so, could you share a pointer?

Even if there is not a good lexer generator utility for Scala, I would think that it would be natural write the lexer using the same parser combinator library we use for parsing. Any particular reason why you didn't go that way?

Reply:
I think the problem is different here. Lexers are turning strings into a stream of tokens. If some strings are not recognizable, a lexer would throw an error. I tried to use the standard lexer approach in AnnotationLexer. It was just too sensitive to the input stream. `CommentPreprocessor` identifies strings that look like annotations and delegates the actual parsing to `AnnotationLexer` and `AnnotationParser`. This approach is less fragile in the presence of arbitrary text.

Reply:
I'm not sure which part of my question that is answering.. :) Could you explain the connections a bit?

Reply:
The short answer is that a lexer generator would solve the problem that is different from our problem. A lexer recognizes good strings and rejects bad strings. A good string is converted into a sequence of tokens. What we need is a preprocessor, that is, a program that receives an arbitrary string, extracts only relevant pieces out of it and passes them to an actual parser. Perhaps, some lexers can do that too, but I have not seen any lexer doing that.

Reply:
Just to note: This seems at odds with your comment in the code. In the comment you say you don't know how to user a lexer generator for this purpose, but here you are saying the problem is that they are not suitable in this case. Is it worth updating the comment to include these insights?

Why we can't deal with bad strings in the lexer via a fallback rule or using a `bad-token` token type? Both approaches are described here: https://stackoverflow.com/a/48896227/1187277

Reply:
True. The comment is a bit outdated. I feel like it should be possible to use a generator for this problem, because I am virtually defining a finite automaton by hand. But since it is not the common case, using a generator for it might happen to be quite hard. I am not sure `bad-token` is exactly solving the problem, but I only had a quick glance.

Reply:
It's a compromise between a framework that one has to learn and 300 LOC that are hard to read. `CommentPreprocessor` is hard to write and read.

Reply:
So my other question is why it wouldn't work to use the same combinator library for the preprocessor, shouldn't that help remove a lot of manual bookkeeping here? Did you find that to be infeasible?

Reply:
Yes, `AnnotationLexer` is using Scala combinators. The previous version tried to parse arbitrary string with it, and it did not work well. I don't know how to a preprocessor with the combinators.

Reply:
Ok. I think we could probably work out a `Parser[String]` for the preprocessing, but we can always do that as a following. The rest of the code here looks fine to me.

Reply:
I don't think that the problem is in picking the library. If you like to put use a lexer generator (or Scala combinators), try to write a state machine first. The edges of this state machine will be regexes. For normal languages, the structure would be something like `(regex1|regex2|...|regexN)`. However, this preprocessor has a non-trivial internal state and it has lots of branching. This is why I do not see how a standard tool will immediately help here.